core/src/main/java/org/infinispan/scattered/ScatteredVersionManager.java

```diff
@@ -0,0 +1,135 @@
+package org.infinispan.scattered;
+
+import org.infinispan.container.versioning.EntryVersion;
+
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * Manages versions of entries and states of segments.
+ *
+ * @author Radim Vansa &lt;rvansa@redhat.com&gt;
+ */
+public interface ScatteredVersionManager<K> {
```

Comment:
I don't think the type is useful and it generates a bunch of warning and casting. Why not using Object as key?